wsl-builder/prepare-build/build.go

```diff
@@ -157,36 +157,62 @@ func downloadRootfses(rootPath string, rootfses string, noChecksum bool) ([]stri
 	return arches, g.Wait()
 }
 
-// downloadFile into dest
+// downloadFile into dest (or copy it if from a local path).
```

Comment:
I don’t really like that a downloadFile function can either copy or download. What do you think about moving this decision more on the caller side (or a function like `isLocalFile(uri string) bool` and deciding where to branch) ?

Reply:
The only one I can see seems not very elegant, replacing every call to downloadFile() by copyOrDownload(), which would be like:
```go
func copyOrDownload(url string, dest string) error {
	if isLocalFile(url){
		return copyLocalFile(url, dest)
	} 		
	
	return downloadFile(url, dest);
}
```
Would that be more idiomatic?

Reply:
Ok, I think in that case we should use a named function for closure we pass to `g.Go()`.
So, this would would become: `getRootfs (uri, rootPath, winArch string, needsChecksum bool) error` which would do:
```
if err := os.MkdirAll(winArch, 0755); err != nil {
    return err
}

if isLocalFile(url)  { // of maybe directly do the stat, if we keep all in the same function
    // probably add a warning that needsChecksum is not supported for local URI if true?
    return copyLocalFile(url, dest)
}

if err := downloadFile(rootfsURL, filepath.Join(rootPath, winArch, "install.tar.gz")); err != nil {
   return err
}

if !needsChecksum {
   return nil
}

u, err := url.Parse(rootfsURL)
if err != nil {
    return err
}
u.Path = filepath.Join(path.Dir(u.Path), "SHA256SUMS")
checksumURL := strings.ReplaceAll(u.String(), "%5C", "/")
checksumDest := filepath.Join(rootPath, winArch, "SHA256SUMS")
if err := downloadFile(checksumURL, checksumDest); err != nil {
    return err
}
if err := checksumMatches(filepath.Join(rootPath, winArch, "install.tar.gz"), filepath.Base(rootfsURL), checksumDest); err != nil {
    return err
}

return nil
```

(Note: I know downloadFile doesn’t follow this, but you don’t need to repeat the type name for arguments when they are the same in following arguments)

What do you think?

Reply:
Agreed. About the checksum, it could be supported but it is very unlikely that a dev or user will edit the SHA256SUMS file to add an entry matching her custom image. (I did that just for experimenting with the previous commit), but with the cost of having the function you proposed more complex, since we couldn't just `return copyLocalFile()` without checksum'ing.

So, I'll proceed with the warning and skip checksum in that case.